utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2112 @@
+  if (handle.offset() <
+      (BlobLogHeader::kHeaderSize + BlobLogRecord::kHeaderSize + key.size())) {
+    if (debug_level_ >= 2) {
+      Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+          "Invalid blob handle file_number: %" PRIu64 " blob_offset: %" PRIu64
+          " blob_size: %" PRIu64 " key: %s",
+          file_number, handle.offset(), handle.size(), key.data());
+    }
+    return Status::NotFound("Blob Not Found, although found in LSM");
+  }
+
+  std::shared_ptr<BlobFile> bfile;
+  {
+    ReadLock l(&mutex_);
+    auto hitr = blob_files_.find(file_number);
+
+    // file was deleted
+    if (hitr == blob_files_.end()) {
+      return Status::NotFound("Blob Not Found as blob file missing");
+    }
+
+    bfile = hitr->second;
+  }
+
+  std::shared_ptr<RandomAccessFileReader> reader;
+  {
+    WriteLock lockbfile_w(&bfile->mutex_);
+    reader = openRandomAccess_locked(bfile, myenv_, env_options_);
+  }
+
+  // allocate the buffer
+  // this is safe in C++11
+  value->resize(handle.size());
+  char* buffer = &(*value)[0];
+
+  Slice blob_value;
+  s = reader->Read(handle.offset(), handle.size(), &blob_value, buffer);
+  if (!s.ok()) {
+    if (debug_level_ >= 2) {
+      Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+          "Failed to read blob from file: %s blob_offset: %" PRIu64
+          " blob_size: %" PRIu64 " key: %s status: '%s'",
+          bfile->PathName().c_str(), handle.offset(), handle.size(), key.data(),
+          s.ToString().c_str());
+    }
+    return Status::NotFound("Blob Not Found as couldnt retrieve CRC");
+  }
+
+  Slice crc_slice;
+  uint32_t crc_exp;
+  std::string crc_str;
+  crc_str.resize(sizeof(uint32_t));
+  char* crc_buffer = &(crc_str[0]);
+  s = reader->Read(handle.offset() - (key.size() + sizeof(uint32_t)),
+                   sizeof(uint32_t), &crc_slice, crc_buffer);
+  if (!s.ok() || !GetFixed32(&crc_slice, &crc_exp)) {
+    if (debug_level_ >= 2) {
+      Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+          "Failed to fetch blob crc file: %s blob_offset: %" PRIu64
+          " blob_size: %" PRIu64 " key: %s status: '%s'",
+          bfile->PathName().c_str(), handle.offset(), handle.size(), key.data(),
+          s.ToString().c_str());
+    }
+    return Status::NotFound("Blob Not Found as couldnt retrieve CRC");
+  }
+
+  uint32_t crc = crc32c::Extend(0, blob_value.data(), blob_value.size());
+  crc = crc32c::Mask(crc);  // Adjust for storage
+  if (crc != crc_exp) {
+    if (debug_level_ >= 2) {
+      Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+          "Blob crc mismatch file: %s blob_offset: %" PRIu64
+          " blob_size: %" PRIu64 " key: %s status: '%s'",
+          bfile->PathName().c_str(), handle.offset(), handle.size(), key.data(),
+          s.ToString().c_str());
+    }
+    return Status::NotFound("Corruption. Blob CRC mismatch");
+  }
+
+  return s;
+}
+
+std::pair<bool, int64_t> BlobDBImpl::sanityCheck(bool aborted) {
```

Comment:
this check is testing system wide sanity of all the files and in-memory metadata.
I think it will be useful for debugging during the first few months of the project.
Once I do periodic sanity checks, if something goes wrong I can know when the
last sanity check succeeded. I can also keep adding recipes to the check as we
discover more bugs/edge cases.